frontend/src/metabase/selectors/metadata.js

```diff
@@ -123,22 +125,69 @@ export const getSegments = createSelector(
+// FIELD VALUES FOR DASHBOARD FILTERS / SQL QUESTION PARAMETERS
+
+// Returns a dictionary of field id:s mapped to matching field values
+// Currently this assumes that you are passing the props of <ParameterValueWidget> which contain the
+// `field_ids` array inside `parameter` prop.
+const getParameterFieldValuesByFieldId = (state, props) => {
+    // NOTE Atte Keinänen 9/14/17: Reading the state directly instead of using `getFields` selector
+    // because `getMetadata` doesn't currently work with fields of public dashboards
+    return _.chain(getIn(state, ["metadata", "fields"]))
+        .pick(getFields(state), ...props.parameter.field_ids)
+        .mapObject(getFieldValues)
+        .value()
+}
 
-    // HACK Atte Keinänen 7/27/17: Currently the field value analysis code only returns a single value for booleans,
-    // this will be addressed in analysis sync refactor
+// Custom equality selector for checking if two field value dictionaries contain same fields and field values
+// Currently we simply check if fields match and the lengths of field value arrays are equal which makes the comparison fast
+// See https://github.com/reactjs/reselect#customize-equalitycheck-for-defaultmemoize
+const createFieldValuesEqualSelector = createSelectorCreator(defaultMemoize, (a, b) => {
+// TODO: Why can't we use plain shallowEqual, i.e. why the field value arrays change very often?
+    return shallowEqual(_.mapObject(a, (values) => values.length), _.mapObject(b, (values) => values.length));
+})
+
+// HACK Atte Keinänen 7/27/17: Currently the field value analysis code only returns a single value for booleans,
+// this will be addressed in analysis sync refactor
+const patchBooleanFieldValues_HACK = (valueArray) => {
     const isBooleanFieldValues =
-        fieldValues && fieldValues.length === 1 && fieldValues[0] && typeof(fieldValues[0][0]) === "boolean"
+        valueArray && valueArray.length === 1 && valueArray[0] && typeof(valueArray[0][0]) === "boolean"
+
     if (isBooleanFieldValues) {
         return [[true], [false]];
     } else {
-        return fieldValues;
+        return valueArray;
     }
 }
 
+// Merges the field values of fields linked to a parameter and removes duplicates
+// We want that we have a distinct selector for each field id combination, and for that reason
+// we export a method that creates a new selector; see
+// https://github.com/reactjs/reselect#sharing-selectors-with-props-across-multiple-components
+// TODO Atte Keinänen 7/20/17: Should we have any thresholds if the count of field values is high or we have many (>2?) fields?
+export const makeGetMergedParameterFieldValues = () => {
+    return createFieldValuesEqualSelector(getParameterFieldValuesByFieldId, (fieldValues) => {
+        const fieldIds = Object.keys(fieldValues)
+
+        if (fieldIds.length === 0) {
+            // If we have no mapped fields, then don't return any values
+            return [];
+        } else if (fieldIds.length === 1) {
```

Comment:
this seems fragile. Couldn't there be other fields that have a single valid field value that aren't booleans?

Reply:
again, it means that there is just a single field, not single field value